### ISSUE TITLE
Proposal: Allow trailing comma in tuples & argument lists

### ISSUE THREAD

**Issue by Neme12** [dotnet/csharplang#1246]
In languages like Pascal the semicolon was a used as a statement separator rather than a terminator, which was a always an annoyance and in my opinion also a design flaw. Languages that have come after have fixed that - but not for commas.

The missing comma causes issues when we put the comma separated list on multiple lines. It shows an extra line as modified in diffs and it makes it harder to reorder the items. The missing comma at the end is always a special case.

I would argue that the idea of a token-separated list rather than token-terminated list is a complicated pattern in the first place and the very definition includes a special case. There is always n-1 separators, unless there is 0 items, in which case there is also 0, not -1 separators. We as programmers know what it's like to write code that joins a string or writes out a comma-separated rather than terminated list. Either the last or the first item is *always* a special case.

I'm not saying you should or will put a comma at the end of your one-line arguments or tuples. I won't. We like to see a comma-separated list surrounded by parentheses. The trailing comma is ugly. But when it gets to multiple lines, it starts getting useful. As our statements get replaced by expressions, setters get replaced by constructor arguments, with tuples and generally C# becoming more expression-oriented, 
commas *are the new semicolon*. Code like this is going to be more and more common:

```c#
return (
    ResourceProvider.Instance.GetElectricityPrice(resources.electricity),
    ResourceProvider.Instance.GetWaterPrice(resources.water),
    ResourceProvider.Instance.GetGasPrice(resources.gas)
    );
```

C# **already allows** trailing comma in enums, array and object initializers. My proposal is to also extend this to all comma-separated parenthesized lists in the language.

**This would also introduce another way to disambiguate a one-tuple expression** (though I would still probably prefer to name the tuple item instead). But it does seem to me like this is the fundamental reason that we can't parse that right now - if we're relying on comma separated lists, *we can't be surprised that there's a special case* for a 1-element tuple.

One disadvantage about is proposal is that it will introduce inconsistencies as some code will include them & some will not (but right now there's the inconsistency that there are already places in the language that allow a trailing comma).

**Comment by yaakov-h** [comment]
Related to #391

**Comment by Neme12** [comment]
The parser would be changed from parsing the first argument and then looping while it sees a comma and expecting another one to looping until it find a close paren, parsing the argument and expecting a comma, for example (with tuples) this:

    if (this.CurrentToken.Kind != SyntaxKind.CloseParenToken)
    {
        var element = ParseTupleElement();
        list.Add(element);

        while (this.CurrentToken.Kind == SyntaxKind.CommaToken)
        {
            var comma = this.EatToken(SyntaxKind.CommaToken);
            list.AddSeparator(comma);

            element = ParseTupleElement();
            list.Add(element);
        }
    }

becomes this:

    while (this.CurrentToken.Kind != SyntaxKind.CloseParenToken)
    {
        var element = ParseTupleElement();
        list.Add(element);

        if (this.CurrentToken.Kind != SyntaxKind.CommaToken)
            break;

        var comma = this.EatToken(SyntaxKind.CommaToken);
        list.AddSeparator(comma);
    }

**Comment by Richiban** [comment]
Unfortunately it conflicts with the proposed syntax for womples, where

```csharp
var x = (1,);
```

Would create an instance of a ValueTuple<int>, i.e. a one-element tuple.

I have to admit though that your proposal for this syntax does fit in better with the rest of C#, where in anonymous object literals, object & collection initializers trailing commas are already allowed.

**Comment by Neme12** [comment]
@Richiban 
How does it conflict? This proposal does the opposite. It would allow that to be parsed as the "womple".

**Comment by softwaremills** [comment]
This is definitely necessary, especially in cases where you're doing something like XLinq and the constructor call would benefit from easy text reordering.

**Comment by theunrepentantgeek** [comment]
At the risk of being flamed for having a contrarian view, I don't believe that _optional_ syntax like this has any place in the language.

Yes, I know that it's already supported in a few places, like object initializers and the like - but compounding an existing mistake by making it worse doesn't improve matters.

Before someone else brings it up - the argument that "If I don't like it, I don't have to use it" falls down as soon as you start looking at long lived codebases with multiple maintainers.

**Comment by softwaremills** [comment]
It's an incredible convenience for enabling reordering of arguments through simple text operations. Being able to select a line and move it up and down is just tremendously useful, especially in a situation like XLinq where parameter lists are being used as collection initializers.

**Comment by CyrusNajmabadi** [comment]
> but compounding an existing mistake 

I'm not sure how it's a mistake.

**Comment by theunrepentantgeek** [comment]
I *really* don't want to come across as a jerk here - while I have a strong opinion on this, there are far more important proposals for the future of C# to discuss. Also, I'm fully aware that there are very smart developers who disagree with me on this - including both the folks who made the decision on adding it to C# as a part of the object initialization syntax, and some of the dev's on my current team.

With all that said ...

*My experience* is that dangling commas correlate very strongly with bugs. When I see one in a code review, or when I'm modifying existing code, it prompts me to look closer. More often than not, I find a problem that's related to the same object. 

Based on that experience, I'd prefer the language not accrue additional syntax that encourages further sloppiness.

**Comment by DavidArno** [comment]
@softwaremills,
> This is definitely necessary...

No, it really isn't. It's arguably useful to those with a certain coding style (and my opinions on that coding style are well documented in the linked issue), but it most certainly isn't necessary.

@CyrusNajmabadi,
>I'm not sure how it's a mistake.

Please reread the duplicate issue, #391 (which you were involved in), and remind yourself of the arguments of those that see trailing commas as a mistake to save us having to repeat the same conversation here.

**Comment by softwaremills** [comment]
@DavidArno You are technically correct, the best kind of correct! Yes, it isn't *necessary*, it just feels like it's necessary to me, and the lack of them shows up as a constant annoyance. It's hugely convenient when *editing* code, while it doesn't matter much when *writing* it.

I appreciate optional trailing commas every time I can Alt-Up or Alt-Down in property initialization without effort, even in simple text editors. Bless you, C# team!

And I curse the lack of optional trailing commas every time I have to cut and paste and fix what is otherwise a completely simple cut and paste between any lines other than the last (or the first, if your commas swing that way). Rearranging lines is an incredibly common activity when editing code. Having syntax accommodate this saves concentration for other things when concentration matters most.

As for readability, when I'm writing CSS, sometimes I write it like this:

    .some-class
    {
        font-size: 110%;
        font-weight: bold;
    }

And sometimes I write it like this:

    .some-class { font-size: 110%; font-weight: bold }

I appreciate CSS letting me put in the optional semicolon in the first place and omitting it in the second case, because in both cases, doing the other thing would hinder readability.

I would agree it's going too far to put optional trailing commas in things like type declarations or parameter declarations. But I would really appreciate optional trailing commas in function call parameter lists.

If your shop really hates optional trailing commas, just have ReSharper flag them as errors in your team policy file.

**Comment by CyrusNajmabadi** [comment]
Yeah.  Looking at the other thread, i come to the same conclusion as before.  Trailing punctuation has more value than problems.  And, C# already recognized this with several constructs *already*.  So it would be sensible to just allow it for all (except the small set of carved out cases where we may want to do something in the language in the future).

**Comment by DavidArno** [comment]
@CyrusNajmabadi,

I'm sure we all come to the same decision as before as this thread has offered nothing to change our minds.

To you, and ~50% of folk that have voted, trailing punctuation has more value than problems. To me, and ~50% folk that have voted, trailing punctuation has more problems than value.

Sure, trailing commas are allowed in some C# constructs already. However, as I said before, "we already made that mistake before, so what harm doing it again" is a pretty rubbish argument. You don't see it as a mistake; I do. So from my point of view, you want to replicate that mistake. And the only reason for wanting to is because it makes it easier to write/edit code. The "consistency" argument is a strawman as everyone excepts there are situations where trailing commas either make no sense or are a breaking change. Making it easy to write code (at the expense of reading it) is a rubbish reason for a change.

You claim it would be sensible to allow it for all lists of values save for a subset where it might affect future plans. But how do you judge those scenarios? This proposal suggests allowing it for tuples. Sounds safe? Nope. consider:
```cs
Foo((1));
Foo((1,));
Foo((1,2));
Foo((1,2,));
```
If trailing commas had been allowed for tuples, then `Foo((1,))` as a one element tuple would be a breaking change. Allowing `Foo((1,2,))` now, but treating  `Foo((1,))` as a one element tuple would be inconsistent. `Foo((1,2,))` is an optional comma;  `Foo((1,))` is mandatory.

Any syntax change risks closing off future ideas. This idea is no different. As you are fond of saying, every idea starts at -100 and has to work upwards. The fact that it risks closing out further ideas and is highly controversial actually sends it more negative. Does "it makes editing a bit easier" really worth all those plus points? Of course it's not.  This should be a dead in the water idea. We have so many way more useful ideas to consider.

**Comment by CyrusNajmabadi** [comment]
As mentioned already: (except the small set of carved out cases where we may want to do something in the language in the future)

**Comment by CyrusNajmabadi** [comment]
>the fact that it risks closing out further ideas and is highly controversial actually sends it more negative

I don't believe this is highly controversial.  In 15+ years i think the amount of feedback we've gotten about trailing commas for existing constructs has been near nil.  Whereas there's been a consistent, though not huge, continual feedback to please just allow this for some really common constructs that just end up being annoying over and over again (namely argument lists).  We've never contemplated any features that this would conflict with, so i'm not really concerned about this impacting some future language change.

--

To me, this just goes into the bucket of cheap thing that just improve language ergonomics.  Most people wouldn't even notice it.  It would just make that common case of working with all sorts of editors much smoother in the common case of reordering things and then having to annoying fix up these syntactic issues.

I'm also unswayed by the argument of "i'm not sure if this was intentional, or a mistake".  For some constructs there are deep and unsettling ambiguities.  And allowing people to write code that really is unclear is certainly a negative for me.  but for trailing commas i don't see any such issue.  

So, it's an easy win in my book.  As i've stated already, i would not take this for scenarios where there may be future language concerns (and where people aren't really asking).  So i wouldn't allow it for type-params, lambda args, or tuples.  But i would certainly take it for argument lists.

**Comment by Neme12** [comment]
@DavidArno 

> Making it easy to write code (at the expense of reading it) is a rubbish reason for a change.

I don't consider it to be harder to read at all. If anything, because of the consistency of having a comma on each line, I'd say it's easier to write *and* read (especially in diffs).

> Foo((1,2,)) is an optional comma; Foo((1,)) is mandatory.

Yes, I'm not too happy about that. It would be inconsistent, but I think that the one element tuple not existing at all is even more inconsistent. How is it a breaking change though?

@softwaremills 

I guess we have a slight disagreement. I would actually put it at the end in CSS even when it's one on line (I didn't know it was valid to omit it). I think looks somewhat less ugly than a trailing comma (when it's on the same line) because of the curlies separated by spaces.
(To me that line actually strikes me as wrong and I can't help but notice the missing semicolon.)

As to parameter definitions, I think if it's allowed on arguments it should definitely be allowed on a parameter list as well for symmetry. And sometimes you do have parameters likes this too, as they showed in the other thread:
@DavidArno 

> Making it easy to write code (at the expense of reading it) is a rubbish reason for a change.

I don't consider it to be harder to read at all. If anything, because of the consistency of having a comma on each line, I'd say it's easier to write *and* read (especially in diffs).

> Foo((1,2,)) is an optional comma; Foo((1,)) is mandatory.

Yes, I'm not too happy about that. It would be inconsistent, but I think that the one element tuple not existing at all is even more inconsistent. How is it a breaking change though?

@softwaremills 

I guess we have a slight disagreement. I would actually put it at the end in CSS even when it's one on line (I didn't know it was valid to omit it). I think looks somewhat less ugly than a trailing comma (when it's on the same line) because of the curlies separated by spaces.
(To me that line actually strikes me as wrong and I can't help but notice the missing semicolon.)

As to parameter definitions, I think if it's allowed on arguments it should definitely be allowed on a parameter list as well for symmetry. And sometimes your parameter list might be formatted like that too, as someone showed before:
```c#
void Foo(
    int bar, 
    int baz,
) { }
```

With tuples I have the same opinion that if it's allowed on tuple expressions, it should be on tuple type literals too for symmetry, even though it's a lot harder to imagine someone splitting them across lines.

**Comment by Neme12** [comment]
@CyrusNajmabadi 

> So i wouldn't allow it for type-params, lambda args, or tuples

What's the problem with tuples?

I have to admit I completely forgot about the (slightly unfortunate) syntax for multidimensional arrays. And I didn't even know about `typeof(Dictionary<,>)`. That does sound concerning to me.

I think I'd still prefer to have it for tuples, arguments & parameters. Definitely not for array indexes or generic type parameters though.

**Comment by softwaremills** [comment]
@Neme12: Yes, the CSS spec is that the semicolon is actually a separator, not a terminator, but it's legal to include the semicolon at the end, probably because it's legal to have empty statements or something.

I can be convinced on normal parameter list declarations. But I think @CyrusNajmabadi's caution on allowing such a thing in lambda parameter list declarations is probably reasonable.

**Comment by Neme12** [comment]
@softwaremills 
Forgive me if I'm missing something but what's the concern with lambdas?

**Comment by softwaremills** [comment]
From issue #391:

> @CyrusNajmabadi:
> Another case i didn't want to change was commas in a lambda. i.e allowing (a, b, c, ) => ... I could imagine us allowing omitted parameters here in the future, and i don't want to make this ambiguous.

**Comment by CyrusNajmabadi** [comment]
> What's the problem with tuples?

It hasn't been certainly decided that we want ```(a,)``` to mean a oneple.  As this decision is still in flight, i would not want to do any work here that impacted ideas there negatively.

**Comment by Neme12** [comment]
@softwaremills 
Hmm, fair enough... I'd definitely be against such a syntax myself as we already have discards (and I don't like the look of it or the idea of counting commas) - but for lambdas underscored probably can't be used for backwards compat (I always preferred asterisks, why did they change it 😕)

**Comment by CyrusNajmabadi** [comment]
The reason to do this feature, for me, is to make it simple to move elements around in a list (i.e. move lines up/down). So the critical cases to handle are those elements that *routinely* cross multiple lines.  Right now, the 99% case for that are parameter lists and argument lists.  So i would start with those.

**Comment by Neme12** [comment]
@CyrusNajmabadi 
Even if the oneple doesn't exist or there is a specialized syntax for it, I don't see how an optional comma would impact that?

I agree it's definitely most useful for argument lists. But it did occur to me after I wrote my first multi-line tuple literal, and I suppose other people will write them too sooner or later.

Yes, I think it still won't be as common as multi-line argument lists, and there will be much lesser need for it since we can't reorder tuples elements as we can do with optional arguments. But it's also true tuples have been designed too look like and be analogous to parameter lists.

**Comment by CyrusNajmabadi** [comment]
> Even if the oneple doesn't exist or there is a specialized syntax for it, I don't see how an optional comma would impact that?

Maybe we will decide that we want ```(a, )``` to mean "a tuple with a default value for the second value".  Now we can't do that because it will now mean "a oneple with the single value 'a'". 

Who knows.  The point is that the space around tuples has not been clearly identified, so we shouldn't blithely take syntactic changes there that might impact future work.

**Comment by Neme12** [comment]
@CyrusNajmabadi 
If that's the case then I don't want it for arguments either. It would be really unexpected if you could add trailing commas in argument lists without any issues but it would change the meaning completely in tuples.

And if you regularly used this syntax for tuples and then saw a trailing comma in an argument list it would also throw you off and you might wonder what it means (if you're used to it having an effect on tuples).

**Comment by CyrusNajmabadi** [comment]
Fair points :)

We may want to wait on this then until we've solidly decided on the syntactic future of tuples.
